src/npctrade.cpp

```diff
@@ -612,50 +663,196 @@ bool trading_window::perform_trade( npc &np, const std::string &deal )
+
+                    for( item_pricing *content : contents ) {
+                        // Only toggle whether item is selected, if its state is different from the container's
+                        // Otherwise deselecting the container, selects the item inside, which makes no sense from UX standpoint
+                        if( content->marked != ip.marked ) {
+                            content->marked = !content->marked;
+                            // make sure we are trading the whole stack
+                            get_change_amount( *content, false );
+                        }
+                        // Handles price calculation, when you've selected few individual items from a container
+                        // and then select the entire container.
+                        // The state that the container is deselected, but some ( not all! ) of it's contents remain selected shouldn't happen
+                        else if( ip.selected && content->selected ) {
+                            // deselect current charge / count
+                            content->selected = false;
+                            content->marked = false;
+                            int change_amount = get_change_amount( *content, false );
+                            adjust_balance( *content, np, change_amount );
+                            // select full charge coutn
+                            content->marked = true;
+                            get_change_amount( *content, false );
+                        }
+                    }
                 }
-                if( ip.loc.where() == item_location::type::character ) {
-                    volume_left += ip.vol * change_amount;
-                    weight_left += ip.weight * change_amount;
+
+                // If we deselected an item inside a container, we also want to deselect the container itself,
+                // cause it would be difficult to handle selling a whole container without one item inside.
+                if( !ip.selected && !ip.marked && ip.parent && ip.parent->marked )
+                {
+
+                    item_pricing *ip_container = &ip;
+
+                    while( ip_container->parent ) {
+                        // Found the top-most selected container
+                        if( ip_container->parent->selected ) {
+                            ip_container = ip_container->parent;
+                            ip_container->selected = false;
+                            ip_container->marked = false;
+
+                            adjust_balance( *ip_container, np, 1 );
+                            break;
+                        } else if( !ip_container->parent->selected && ip_container->parent->marked ) {
+                            ip_container = ip_container->parent;
+                            ip_container->marked = false;
+                        } else {
+                            break;
+                        }
+                    }
```

Comment:
i feel like this is ad-hoc code of something we already have in the code base somewhere, but i can't remember what it is off-hand.

Reply:
my first suspicion is that this is just a reimplementation of the parent-finding code in item_location.